Build tuid generator

ajupazhamayil opened the pull request:
#115

ajupazhamayil committed:
init_es in clogger

ajupazhamayil committed:
made CSETLOG_SCHEMA constant

ajupazhamayil committed:
delete and init is not working properly

ajupazhamayil committed:
dirty commit with ES queries

ajupazhamayil committed:
Merge dev and make clogger work with ES

ajupazhamayil committed:
testing

ajupazhamayil committed:
Sometimes tests are failing

ajupazhamayil committed:
working with test_basic errors

ajupazhamayil committed:
resolving test_basic errors

ajupazhamayil committed:
passed all tests: used refresh and lock

ajupazhamayil committed:
removes unnecessary transactions, changed tests order

ajupazhamayil committed:
format using python black

ajupazhamayil committed:
changes function return type

ajupazhamayil committed:
tests alligns with no recomputation

ajupazhamayil committed:
Merge branch 'dev' into elasticsearch

ajupazhamayil committed:
initial commit

ajupazhamayil committed:
test_basic and test_clogger works

ajupazhamayil committed:
changes test_memory, could not test

ajupazhamayil committed:
done the changes requested

ajupazhamayil committed:
initial commit for a new tuid generator

ajupazhamayil commented on tuid/service.py:
```diff
@@ -1499,17 +1501,32 @@ def get_tuids(self, files, revision, commit=True, chunk=50, repo=None):
+                        {"eq": {"annotated_files": annotated_files}},
+                        {"in": {"path": annotations_to_get}},
+                        {"eq": {"branch": repo}}
+                    ]},
```
@klahnakoski I had to add 2 more "eq" to make the request. Kindly let me know how to restructure it such that the request same as the regular tuid request.

klahnakoski replied:
`annotated_files` is another word for `path`  and `revision` is another word for `commit`.  I think you are using different names for the same thing.

klahnakoski commented:
@ajupazhamayil   It is not easy to follow what you have done here.  You do not need those extra properties: The processes calling the tuid generator should not tell the tuid generator anything because they can not be trusted to know what the tuids are.  The tuid generator should go to the ES cluster (and its memory cache) to determine what tuids are missing, and fill those in.

ajupazhamayil commented:
@klahnakoski I was actually trying to take out this function https://github.com/mozilla/TUID/blob/faeb78848d46f0afe108579262f552768b8d956a/tuid/service.py#L1647 and make it as a service. This function checks for missing tuids in a file, fill it and returns the result. Am I correct?

gmierz commented:
@ajupazhamayil No, you only need to handle TUID generation. The _get_tuids function does more than just generate tuids. See @klahnakoski comment for more information: "The tuid generator should go to the ES cluster (and its memory cache) to determine what tuids are missing, and fill those in." 

The service will be accepting a single file and revision, along with the line numbers. These line numbers might already have a TUID which you will need to check for.

ajupazhamayil commented:
@gmierz @klahnakoski I'm a little bit confused here. Are we looking for something like this?

```
def tuid_generator( self, file, revision, line_nums):
    results = {}
    completed = True
    with self.conn.transaction() as transaction:
        tmp_ann = self._get_annotation(revision, file, transaction=transaction)
        tuids = {}
        for line in tmp_ann.split("\n"):
            tuid, line_num = line.split(",")
            tuids[line_num] = tuid
        if not tuids:
            print("file in that revision does not exist in the DB")
            completed = False
            return {}, completed

        for line in line_nums:
            if line in tuids:
                results[line] = tuids[line]
            else:
                results[line] = self.tuid()

    return results, completed

```

gmierz commented:
Yes that's right, except you don't need to get the annotation.

klahnakoski commented:
@ajupazhamayil @gmierz 

There are two ways to do this:

1) Have the process apply the changesets to the files and then ask the tuid generator to fill in the holes.  To do this, we must pass in the known tuids, and the number of lines in the file.  
2) In the event that the tuids are not all in ES, then ask the tuid generator for (file, revision);  The tuid generator will apply any changesets and fill in the missing tuids.

I was in thinking of the latter; The tuid generator would have the same API as the tuid service does now.  But, it also demands that the changeset application be done by the tuid generator.  This is why I was worried about there still being a bottleneck. 

If we do the former, then the changeset application is distributed, maybe redundant, but faster.  What can we say about the `line_nums` passed to the tuid generator?  Must we make a tuid generator request for every revision along a changeset path?  I imagine one process applying two changesets, then asking for missing tuids of the result, while a another process only applies one changeset and asking for a result can cause tuid collisions. Are other processes guaranteed to pass the same (file, revision, line_nums)?  What if the changesets are applied backward? Do merge conflicts cause a problem?  Is there such thing as a flle frontier when multiple processes manage the frontiers?

I still prefer the latter because the information flows in one direction; out from the tuid generator.  While the latter might work, but seems fraught with coordination issues.

ajupazhamayil commented:
@klahnakoski @gmierz My doubt is, If we do the 2nd option, Almost all the time number of processes working will be only one right? Because all the other process has to wait until this full generation process completes.

klahnakoski commented:
@ajupazhamayil   yes, you got it.

klahnakoski commented:
@ajupazhamayil as per the discussion #codecoverage: Sorry, this PR was a bad idea.

see https://github.com/mozilla/TUID/issues/118
